applications: nrf_desktop: Add BLE advertising documentation (#2079)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/ble_adv.rst

> BLE advertising module

Author: I think we should leave it as it is. Other application modules are named using BLE abbreviation.

######################

Use the BLE advertising module to control the Bluetooth LE advertising.

Module Events
*************

+------------------+------------------------------+-------------+---------------------------+-------------------+
| Source Module    | Input Event                  | This Module | Output Event              | Sink Module       |
+==================+==============================+=============+===========================+===================+
| :ref:`ble_state` | ``ble_peer_event``           | ``ble_adv`` |                           |                   |
+------------------+------------------------------+             |                           |                   |
| :ref:`ble_bond`  | ``ble_peer_operation_event`` |             |                           |                   |
+------------------+------------------------------+             +---------------------------+-------------------+
|                  |                              |             | ``ble_peer_search_event`` | :ref:`led_state`  |
+------------------+------------------------------+-------------+---------------------------+-------------------+

> Configuration

Author: Ok, I will make a list of steps from mandatory elements.

*************

> The module requires the basic Bluetooth configuration, as described in the Bluetooth guide.

Author: The guide is not merged yet. We need to merge it before we add link here.

Set the ``CONFIG_DESKTOP_BLE_ADVERTISING_ENABLE`` to enable the BLE advertising module for the nRF Desktop peripheral device.

You must define the short device name used for advertising (``CONFIG_DESKTOP_BLE_SHORT_NAME``), the Bluetooth device name (:option:`CONFIG_BT_DEVICE_NAME`) and the Bluetooth device appearance (:option:`CONFIG_BT_DEVICE_APPEARANCE`).

By default, the module uses indirect advertising.

> Set the ``CONFIG_DESKTOP_BLE_DIRECT_ADV`` option to use direct advertising.

Author: I think both are proper.
https://www.bluetooth.com/blog/advertising-works-part-1/

> Set the ``CONFIG_DESKTOP_BLE_DIRECT_ADV`` option to use direct advertising.

Author: I will add this.

> .. tip::

Author: ok

   The nRF Desktop peripheral will not advertise directly towards central that uses Resolvable Private Address.

>    The peripheral does not check if the remote device supports address resolution of directed advertisements - it does not read the Central Address Resolution GATT characteristic of the central.

Author: I am not sure if because is a good word here.
You check if the remote device supports the address resolution of directed advertisements by reading the Central Address Resolution GATT characteristic of the central

>    The peripheral does not check if the remote device supports address resolution of directed advertisements - it does not read the Central Address Resolution GATT characteristic of the central.

Author: I will update this.

Set the ``CONFIG_DESKTOP_BLE_FAST_ADV`` to make the peripheral initially advertise with shorter interval to enquicken finding the peripheral by Bluetooth centrals.

> If the device uses indirect advertising, it will switch to slower advertising after the ``CONFIG_DESKTOP_BLE_FAST_ADV_TIMEOUT`` seconds.

Author: ok

> If the device uses direct advertising, you will observe the ``ble_peer_event`` with :cpp:member:`state` set to :cpp:enumerator:`PEER_STATE_CONN_FAILED` in case the Bluetooth central does not connect during predefined period of fast direct advertising.

Author: Then obviously the advertising is stopped. The directed advertising is directed only to one specified central. If its already connected there is no reason to advertise towards it.

> After the event is received, the device will switch to low duty direct advertising.

Author: The advertising is interrupted on the CONN_FAILED. The `switch to` expression was better here.

Switching to slower advertising is done to reduce the energy consumption.

You can use the ``CONFIG_DESKTOP_BLE_SWIFT_PAIR`` option to enable the Swift Pair feature.
The feature simplifies pairing the Bluetooth peripheral to Windows10 hosts.

When system goes to power down state, the advertising is stopped.
If the Swift Pair was enabled, the device advertises without the Swift Pair data for additional ``CONFIG_DESKTOP_BLE_SWIFT_PAIR_GRACE_PERIOD`` seconds to ensure that user will not try to connect to the device that is no longer available.

> Implementation details

Author: This is more common issue for nrf_desktop documentation. I think we should discuss this.

**********************

> If the currently used Bluetooth local identity already has a bond and the device uses indirect advertising, it will not set General Discoverable flag.

Author: the advertising device
I will add this information.

> If the :option:`CONFIG_BT_WHITELIST` is enabled, the device will also whitelist incoming scan response data requests and connection requests.

Reviewer: Is this related to the flag mentioned in the sentence above?

Author: It has similar purpose, but these are two separate actions.

This is done to prevent Bluetooth centrals other than the bonded one from connecting with the device.

> The nRF Desktop dongle scans for peripheral devices using part of the scan response data - Bluetooth device name.

Reviewer: ```suggestion
The nRF Desktop dongle scans for peripheral devices using the Bluetooth device name, which is provided in the scan response data.
```

> The BLE advertising module uses the :ref:`zephyr:settings` to store the information if the peer for given local identity uses RPA address.

Reviewer: ```suggestion
The BLE advertising module uses the :ref:`zephyr:settings` to store the information if the peer for the given local identity uses the Resolvable Private Address (RPA).
```
